tests/integration/generate_relationships.py

```diff
@@ -0,0 +1,496 @@
+#!/usr/bin/env python3
+"""
+Generate expected entity state events and relationships JSON files from ClickHouse data.
+
+This script connects to ClickHouse (via port-forward) and generates JSON files
+for integration test assertions. It automatically discovers all entity types and
+relationship types in the database.
+
+Prerequisites:
+    pip install clickhouse-connect
+
+Usage:
+    # Make sure Skaffold is running (handles port-forwarding automatically):
+    skaffold dev
+
+    # Run this script to generate all discovered entities and relationships:
+    python3 generate_relationships.py
+
+    # Or with custom host/port:
+    python3 generate_relationships.py --host localhost --port 8123
+"""
+
+import argparse
+import json
+import os
+import re
+import sys
+from typing import Dict, List, Any
+
+try:
+    import clickhouse_connect
+except ImportError:
+    print("Error: clickhouse-connect not installed", file=sys.stderr)
+    print("Install it with: pip install clickhouse-connect", file=sys.stderr)
+    sys.exit(1)
+
+
+# Constants
+TEST_PREFIX = 'test-'
+CLUSTER_UID_KEY = 'sw.k8s.cluster.uid'
+POD_NAME_KEY = 'k8s.pod.name'
+REPLICASET_NAME_KEY = 'k8s.replicaset.name'
+
+# Kubernetes workload name keys in priority order
+K8S_WORKLOAD_KEYS = [
+    'k8s.deployment.name',
+    'k8s.statefulset.name',
+    'k8s.daemonset.name',
+    'k8s.service.name',
+    'k8s.pod.name',
+    'k8s.replicaset.name',
+    'k8s.job.name'
+]
+
+# Hash pattern constants
+POD_HASH_LENGTH = 5
+REPLICASET_HASH_MIN_LENGTH = 8
+REPLICASET_HASH_MAX_LENGTH = 10
+
+
+def parse_entity_id(entity_id_str: str) -> Dict[str, str]:
+    try:
+        return json.loads(entity_id_str)
+    except (json.JSONDecodeError, TypeError):
+        return {}
+
+
+def to_snake_case(name: str) -> str:
+    s1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
+    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).lower()
+
+
+def get_workload_name(entity_id: Dict[str, str]) -> str:
+    for key in K8S_WORKLOAD_KEYS:
+        if key in entity_id:
+            return entity_id[key]
+    return ""
+
+
+def _check_pod_hash_pattern(pod_name: str) -> bool:
+    """Returns True if pod name matches: <base>-<8-10chars>-<5chars> pattern.
```

Comment:
The actual logic can be more complicated. E.g. if the `<base>-<8-10chars>-` is too long, the resulting Pod name will end with `<5chars>` preceded by `<base>-<8-10chars>-` trimmed from the right to fit.